### PR TITLE
Major overhaul and tree converting

### PR DESCRIPTION
**Pull request by BenjaBobs**
See #38 for a description.

### CODE REVIEW

File: src/Gremlin.Net.CosmosDb/Serialization/AutoConnector.cs [hunk 1]
```diff
@@ -0,0 +1,118 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Gremlin.Net.CosmosDb.Structure;
+
+namespace Gremlin.Net.CosmosDb.Serialization
+{
+    internal class AutoConnector : IVertexConnector, IEdgeConnector
```
Comment (evo-terren) [inline src/Gremlin.Net.CosmosDb/Serialization/AutoConnector.cs]: Can you do me a favor and add some comments throughout this class and methods just to help future devs?

File: src/Gremlin.Net.CosmosDb/Serialization/AutoConnector.cs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            {
+                property.SetValue(vertex, edge);
+            }
+            else
+            {
+                var vertexProps = edge.GetType().GetProperties().Where(p => TypeCache.IVertex.IsAssignableFrom(TypeHelper.UnderlyingType(p.PropertyType)));
+
+                foreach (var prop in vertexProps)
+                {
+                    if (prop.GetValue(edge) is object newEdgeVertex)
```
Comment (evo-terren) [inline src/Gremlin.Net.CosmosDb/Serialization/AutoConnector.cs]: NIT: Using a switch here may be a bit cleaner than all the if/is checks you're doing:
```c#
switch (prop.GetValue(edge))
{
    case IVertex v:
        ConnectVertex(existingEdge, v);
        break;
    case IEnumerable vEnum:
        foreach (var vert in vEnum.OfType<IVertex>())
        {
            ConnectVertex(existingEdge, vert);
        }
        break;
}
```

File: src/Gremlin.Net.CosmosDb/Structure/ManyToManyEdge.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    /// <typeparam name="TOutV">The Out vertex type</typeparam>
+    /// <typeparam name="TInV">The In vertex type</typeparam>
+    public class ManyToManyEdge<TOutV, TInV> : IEdge<TOutV, TInV>
+        where TOutV : IVertex
+        where TInV : IVertex
+    {
+        /// <summary>
+        /// The list of Out vertices.
+        /// </summary>
+        public List<TOutV> OutV { get; set; }
```
Comment (evo-terren) [inline src/Gremlin.Net.CosmosDb/Structure/ManyToManyEdge.cs]: This should probably be `IReadOnlyCollection<TOutV>` or at least `IList<TOutV>`.

File: src/Gremlin.Net.CosmosDb/Structure/ManyToManyEdge.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+        /// <summary>
+        /// The list of In vertices.
+        /// </summary>
+        public List<TInV> InV { get; set; }
```
Comment (evo-terren) [inline src/Gremlin.Net.CosmosDb/Structure/ManyToManyEdge.cs]: Same here

File: src/Gremlin.Net.CosmosDb/Structure/ManyToOneEdge.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    /// <typeparam name="TOutV">The Out vertex type</typeparam>
+    /// <typeparam name="TInV">The In vertex type</typeparam>
+    public class ManyToOneEdge<TOutV, TInV> : IEdge<TOutV, TInV>
+        where TOutV : IVertex
+        where TInV : IVertex
+    {
+        /// <summary>
+        /// The list of Out vertices.
+        /// </summary>
+        public List<TOutV> OutV { get; set; }
```
Comment (evo-terren) [inline src/Gremlin.Net.CosmosDb/Structure/ManyToOneEdge.cs]: And here

File: src/Gremlin.Net.CosmosDb/Structure/OneToManyEdge.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        /// <summary>
+        /// The Out vertex.
+        /// </summary>
+        public TOutV OutV { get; set; }
+
+        /// <summary>
+        /// The list of In vertices.
+        /// </summary>
+        public List<TInV> InV { get; set; }
```
Comment (evo-terren) [inline src/Gremlin.Net.CosmosDb/Structure/OneToManyEdge.cs]: You get the idea...

File: src/Gremlin.Net.CosmosDb/Structure/Tree.cs [hunk 7]
```diff
@@ -10,5 +16,91 @@ public class Tree
...
         /// </summary>
         public TreeVertexNode[] RootVertexNodes { get; set; }
+
+        /// <summary>
+        /// Convert the parsed tree into the specified type.
+        /// </summary>
+        /// <typeparam name="T">The type to convert to.</typeparam>
+        /// <param name="treeConnectors">Custom connectors to use for attaching edges and vertices.</param>
+        /// <returns></returns>
+        public T[] ToObject<T>(List<ITreeConnector> treeConnectors = null)
```
Comment (evo-terren) [inline src/Gremlin.Net.CosmosDb/Structure/Tree.cs]: Can you please change this signature to be the following

```c#
public IEnumerable<T> ToObject<T>(IEnumerable<ITreeConnector> treeConnectors = null)
```

**Comment by BenjaBobs** [comment]
I addressed the pr issues and found (and fixed) a small error.

**Pull request merged**